Fix creating a quiz without having imported resources or users enrolled in the class

AllanOXDi opened the pull request:
## Summary

This pull request adds warning messages for scenarios in which there are no imported channels and/or users associated with the class when a coach attempts to create a quiz.

## References
[#12330](https://github.com/learningequality/kolibri/issues/12330)
## Reviewer guidance

1. Please review the slack conversations for some decisions made whether to use a model or not [here ](https://learningequality.slack.com/archives/CB37UM23A/p1729271450496829)and[ here](https://learningequality.slack.com/archives/CTG8XDTCL/p1729629998736479)
2. Does the changes makes sense ?

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/CoachExamsPage/index.vue:
```diff
@@ -7,6 +7,29 @@
+          </div>
+
+          <div class="error-message">
+            <p v-if="channels.length === 0">{{ noResourcesAvailable$() }}</p>
```
[improvement, non-blocking] To simplify this, you could put the condition on message container div here, rather than on the `<p>` tag and on the link separately. (But keep the `v-if="deviceContentUrl` on the link)

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/CoachExamsPage/index.vue:
```diff
@@ -353,8 +385,25 @@
+        return '';
+      },
+      hasNoChannels() {
+        return !this.channels || this.channels.length === 0;
```
[blocking, small cleanup] Doing this type of check is typically a good idea when fetching data, but it's unnecessary here. Since you have set up `channels` in `data` to be an empty array, `this.channels` will always exist, even if the channels haven't loaded yet. That means, the critical piece here is the "`this.channels.length === 0`"

[improvement, non-blocking] In your template code, you are using `channels.length === 0` on the banner and then you use this `hasNoChannels` computed prop for the button. I think it would be easier to see when reading the code that these are always "opposite" conditions, meaning that if one shows up, the other doesn't, if your code is more clearly referencing the same condition in both places. 

You could either update this computed prop to just be checking if `this.channels.length` and not worry about the existence check, and use `hasNoChannels` in both places, or you could delete this computed prop entirely and just use a length check in both places.

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/CoachExamsPage/index.vue:
```diff
@@ -408,6 +457,17 @@
+          getParams: {
+            contains_exercise: true,
+            available: true,
+            contains_quiz: true,
```
[blocking] `contains_quiz` is not the most intuitive, because this isn't actually about for being able to create a quiz, but rather, whether or not a channel contains practice quizzes. Including this as `true` therefore is more restriction than you want. 

`contains_exercise: true` is definitely always required for a channel to be able to be used to create a quiz, but `contains_quiz` is not required. 

To see more specifically what I mean, you can use the Kolibri QA channel. Importing a few exercises, but no practice quizzes means that this fetch will return an empty array.

https://github.com/user-attachments/assets/15396efc-8336-4b38-baa4-44db4cd39334

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/assignments/IndividualLearnerSelector.vue:
```diff
@@ -32,7 +32,7 @@
           <CoreTable
             :selectable="true"
-            :emptyMessage="$tr('noUsersMatch')"
+            :emptyMessage="noLearnersEnrolled$({ className: className })"
```
[blocking] what happens here if you do have learners [learner_a, learner_b] enrolled in the class, and you search in the learners table for "learner_c"?

<img width="975" alt="Screenshot 2024-10-24 at 4 07 24 PM" src="https://github.com/user-attachments/assets/422a91e4-4300-4834-a99f-4695e222b1d0">

AllanOXDi replied:
I miss this. I think we will still see `noLearnerEnrolled` string. Which does not make sense. Should  we haddled that with` noUsersMatch' string?

marcellamaki replied:
Yes, I think handling that with the `noUsersMatch` string, and using one or the other based on the condition, would be the way to go

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/CoachExamsPage/index.vue:
```diff
@@ -353,8 +385,25 @@
       },
+      deviceContentUrl() {
+        const deviceContentUrl = urls['kolibri:kolibri.plugins.device:device_management'];
+        if (deviceContentUrl && this.canManageContent) {
```
nice work with this!

marcellamaki reviewed (dismissed):
Hi @AllanOXDi - overall, this is very close to being done, and the changes I'm requesting hopefully should be more of a little bit of tidying up than anything. 

[blocking] This PR deletes the `yarn.lock` file which I think is maybe an unintended side effect of your rebasing. This is also causing a problem with the checks - you can see there's an issue in the dependencies that's preventing the whl from building which prevents the other build assets from running as well. If you update this, hopefully that should resolve the problem! If after you update this, the checks are still failing, let me know so we can troubleshoot further.

[improvement, non-blocking] can you think of a way to add some `loading` or state management logic to help prevent the potential for a flash of the banner? Not required - there are a lot of pre-existing loading state challenges in Coach.

https://github.com/user-attachments/assets/39160cbe-2d43-44a2-9d55-b3568604aa7c

pcenov reviewed (approved):
Hi @AllanOXDi, on my end I confirm that the 2 issues described in https://github.com/learningequality/kolibri/issues/12330 are now fixed so if there's nothing else in terms of code cleanup this should be good to go!

marcellamaki commented:
Hi @AllanOXDi - this is ready to go, but can you please remove the `yarn.lock` file as it is not related to this changes? Then we can merge

AllanOXDi committed:
adds no learners enrolled string
disabled the add quiz button
do not show routerlink link when there is no resource

rtibbles reviewed (approved):
Have done a small rebase to get this mergeable. Confirmed that the fix is still in place after rebase.
